cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -563,28 +582,60 @@ private ConcurrencyAdjustingRecommendation minIsrBasedConcurrency() {
       return ExecutionUtils.recommendedConcurrency(cluster, minIsrWithTimeByTopic);
     }
 
-    private void decreaseExecutionConcurrencyForBroker(int brokerId, ConcurrencyType concurrencyType) {
+    private void decreaseExecutionBrokerConcurrency(int brokerId, ConcurrencyType concurrencyType) {
+      int currentMovementConcurrency = _executionConcurrencyManager.getExecutionBrokerConcurrency(brokerId, concurrencyType);
+      int decreasedConcurrency = getDecreasedConcurrency(currentMovementConcurrency, concurrencyType);
+      if (decreasedConcurrency > 0) {
+        _executionConcurrencyManager.setExecutionConcurrencyForBroker(brokerId, decreasedConcurrency, concurrencyType);
+        LOG.info("Concurrency adjuster decreased the {} movement concurrency to {} for broker {}.",
+            concurrencyType, decreasedConcurrency, brokerId);
+      }
+    }
+
+    private void decreaseExecutionClusterConcurrency(ConcurrencyType concurrencyType) {
+      int currentMovementConcurrency = _executionConcurrencyManager.getExecutionClusterConcurrency(concurrencyType);
+      int decreasedConcurrency = getDecreasedConcurrency(currentMovementConcurrency, concurrencyType);
+      if (decreasedConcurrency > 0) {
+        _executionConcurrencyManager.setExecutionConcurrencyForAllBrokers(decreasedConcurrency, concurrencyType);
+        LOG.info("Concurrency adjuster decreased the {} movement concurrency to {}.", concurrencyType, decreasedConcurrency);
+      }
+    }
+
+    private int getDecreasedConcurrency(int currentConcurrency, ConcurrencyType concurrencyType) {
       int minMovementsConcurrency = MIN_CONCURRENCY.get(concurrencyType);
-      int currentMovementConcurrency = _executionConcurrencyManager.getExecutionConcurrency(brokerId, concurrencyType);
-      // Multiplicative-decrease reassignment concurrency (MIN: minMovementsConcurrency).
-      if (currentMovementConcurrency > minMovementsConcurrency) {
-        int recommendedConcurrency = Math.max(minMovementsConcurrency, currentMovementConcurrency / MULTIPLICATIVE_DECREASE.get(concurrencyType));
-        _executionConcurrencyManager.setExecutionConcurrencyForBroker(brokerId, recommendedConcurrency, concurrencyType);
-        LOG.info("Concurrency adjuster decreased the {} movement concurrency to {} for broker {}",
-                 concurrencyType, recommendedConcurrency, brokerId);
+      if (currentConcurrency > minMovementsConcurrency) {
+        // Multiplicative-decrease reassignment concurrency (MIN: minMovementsConcurrency).
+        return Math.max(minMovementsConcurrency, currentConcurrency / MULTIPLICATIVE_DECREASE.get(concurrencyType));
       }
+      return -1;
```

Comment:
Why not just return minMovementConcurrency here so that you don't need to check if "(decreasedConcurrency > 0) " in L598?

Reply:
Thought more on this, actually, if the concurrency is explicitly set to a small value through rest api (< minMovementsConcurrency), we should let it as. So change it return currentConcurrency if no need change.